Source/Dafny/Resolver.cs

```diff
@@ -14822,9 +14816,16 @@ public class ResolveOpts
       } else if (ctor == null) {
         reporter.Error(MessageSource.Resolver, pat.tok, "constructor {0} does not exist in datatype {1}", pat.Id, dtd.Name);
       } else {
-        var argCount = pat.Arguments == null ? 0 : pat.Arguments.Count;
-        if (ctor.Formals.Count != argCount) {
-          reporter.Error(MessageSource.Resolver, pat.tok, "pattern for constructor {0} has wrong number of formals (found {1}, expected {2})", pat.Id, argCount, ctor.Formals.Count);
+        if (pat.Arguments == null) {
+          if (ctor.Formals.Count == 0) {
+            // The Id matches a constructor of the correct type and 0 arguments,
+            // so make it a nullary constructor, not a variable
+            pat.MakeAConstructor();
+          }
+        } else {
+          if (ctor.Formals.Count != pat.Arguments.Count) {
```

Comment:
This could potentially be an `else if` instead (saving one level of block statements in the code).

Reply:
OK